### ISSUE TITLE
[HOLD for payment 2023-05-16] [HOLD for payment 2023-05-09] [$1000] Add Phone Number validation to New Chat field

### ISSUE THREAD

**Issue by stitesExpensify** [Expensify/App#17461]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the floating action menu
2. Select "New Chat"
3. Type in an invalid phone number IE `+444`
4. Create the chat
5. You will get an error

## Expected Result:
We should validate that the entered phone number is in a valid format. We want to make sure that entered phone numbers follow the correct pattern for a given country code, regardless of formatting i.e. (+1-123-123-1234, +11231231234, +1.123.123.1234) should all work. Preferably this would be done with Regex, but other proposals are welcome. Additionally we need to ensure that this is done for all search types (new chat, group chat, split bill, etc).

## Actual Result:
You can create the chat with an obviously broken phone number

## Workaround:
Just don't enter those numbers

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:**
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:** n/a
**Email or phone of affected tester (no customers):** All
**Logs:** https://stackoverflow.com/c/expensify/questions/4856 n/a
**Notes/Photos/Videos:** Any additional supporting documentation 
**Expensify/Expensify Issue URL:**
**Issue reported by:** @stitesExpensify 
**Slack conversation:** https://expensify.slack.com/archives/C03U7DCU4/p1681418937228539?thread_ts=1680815598.330879&cid=C03U7DCU4

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01319c46dbb72458cd</li>
        <li>Upwork Job ID: 1646928208403300352</li>
        <li>Last Price Increase: 2023-04-14</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01319c46dbb72458cd

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Triggered auto assignment to @maddylewis (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @s77rt (`External`)

**Comment by MelvinBot** [comment]
Current assignee @stitesExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by s77rt** [comment]
@Prince-Mendiratta Thanks for the proposal. I think we want to use [awesome-phonenumber](https://github.com/grantila/awesome-phonenumber) because it's something already being added here https://github.com/Expensify/App/pull/16804. Can you please update your proposal based on that?

Also please make sure to cover the following cases:
1. Validation of international phone numbers (phone numbers with country code)
2. Validation of local phone numbers

**Comment by apt-get-sahil** [comment]
**Proposal****
Please re-state the problem that we are trying to solve in this issue.**
Add Phone Number validation to New Chat field 
**What is the root cause of that problem?**
This is a feature request, not a bug.
**What changes do you think we should make in order to solve the problem?**

Using regular expressions (regex) to validate phone numbers can help solve the error in Expensify where an invalid phone number can be entered in the "New Chat" feature.

The regex pattern (^\+(?:[0-9]●?){6,14}[0-9]$) checks whether a string starts with a plus sign (+) and is followed by between 6 and 14 digits (inclusive). If the input matches this pattern, the phone number is considered valid.

By using this regex pattern to validate phone numbers before creating a chat, we can prevent invalid phone numbers from being submitted and causing errors. This ensures that the phone number entered by the user is in a valid format, which will help prevent any issues with sending messages to that phone number later on.

**What alternative solutions did you explore? (Optional)**
Using the awesome-phonenumber library to validate phone numbers can also help solve the error in Expensify's "New Chat" feature.

**Comment by MelvinBot** [comment]
📣 @apt-get-sahil! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by apt-get-sahil** [comment]
Contributor details
Your Expensify account email: sahilnagel@gmail.com
Upwork Profile Link: upwork.com/freelancers/~0118a80dbcb14a0994

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by alitoshmatov** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Add Phone Number validation to New Chat field
### What is the root cause of that problem?
Currently we are not validating phone numbers
### What changes do you think we should make in order to solve the problem?
As mentioned above, we should use [awesome-phonenumber](https://github.com/grantila/awesome-phonenumber) package to validate phone numbers. 

We can make use of `parsePhoneNumber` method and check if this phone number is possible or not. If yes we can get pure numbers from `e164` field
<img width="609" alt="Screenshot 2023-04-15 at 2 40 06 PM" src="https://user-images.githubusercontent.com/59907218/232206146-6565ffeb-295c-42bc-af92-4c859367d762.png">

Regarding your second point validation local phone numbers is not supported in this package, and this is good as omitting country code might result in overlapping phone numbers. Accounts also required to register phone number with country code too. 


### What alternative solutions did you explore? (Optional)

**Comment by s77rt** [comment]
@apt-get-sahil Thanks for the proposal. Your solution is only good to check if the input looks like a number or not but it does not validate it e.g. `+11111111` should not pass.

**Comment by s77rt** [comment]
@alitoshmatov Thanks for the proposal. I think your suggested solution is not complete for not taking into consideration local numbers.

**Comment by s77rt** [comment]
@Prince-Mendiratta Thanks for the update. I think your solution is too complex and contains repeated and redundant logic. For instance adding country code to phone numbers can be done using [appendCountryCode](https://github.com/Expensify/App/blob/4607dcb69b6d5bb7607f652f8a78327c7e002183/src/libs/LoginUtils.js#L38). Also I don't see why you are making changes to the header message, if the number is not valid it should not be shown as an option to be selected in the first place.

**Comment by s77rt** [comment]
Using [awesome-phonenumber](https://github.com/grantila/awesome-phonenumber) I think we can achieve the wanted results with a single line of code.

**Comment by s77rt** [comment]
@Prince-Mendiratta Thanks for the update. Now it's much simpler, but lets only show valid numbers.
```diff
-        && ((Str.isValidEmail(login) && !Str.isDomainEmail(login)) || Str.isValidPhone(login))
+        && ((Str.isValidEmail(login) && !Str.isDomainEmail(login)) || parsePhoneNumber(login).valid)
```
This should be enough but let's discuss further in the PR.

:ribbon: :eyes: :ribbon: C+ reviewed

cc @stitesExpensify

**Comment by s77rt** [comment]
@Prince-Mendiratta We may update the validation on header message too (switch from  `Str.isValidPhone` to `parsePhoneNumber`)

I have raised a Slack thread here https://expensify.slack.com/archives/C01GTK53T8Q/p1681596380231529

cc @stitesExpensify

**Comment by s77rt** [comment]
@rushatgabhane It's hard to say yes XD Can you please provide more context why this should be put on hold?

**Comment by s77rt** [comment]
If we are still going to use [awesome-phonenumber](https://github.com/grantila/awesome-phonenumber) then there is no reason to hold as the lib does not care about formatting.

**Comment by puneetlath** [comment]
I think actually the PRs can go in either order. https://github.com/Expensify/App/pull/16804 uses [awesome-phonenumber](https://github.com/grantila/awesome-phonenumber) for displaying phone numbers. Whereas this issue will use [awesome-phonenumber](https://github.com/grantila/awesome-phonenumber) for validating them. So I think we can go forward in parallel.

**Comment by s77rt** [comment]
@rushatgabhane For consistency I think that would make sense.

**Comment by maddylewis** [comment]
making sure @s77rt and @rushatgabhane addressed this comment here - https://github.com/Expensify/App/issues/17461#issuecomment-1513206227

i think so! but wanted to make sure https://github.com/Expensify/App/pull/17578#issuecomment-1515909865

**Comment by trjExpensify** [comment]
Not sure why two BZ got assigned to this issue. @maddylewis you've been involved more actively to date, so I'll drop off. 👍

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.8-8 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17578


If no regressions arise, payment will be issued on 2023-05-09. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@s77rt] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@s77rt] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@s77rt] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@s77rt] Determine if we should create a regression test for this bug.
- [ ] [@s77rt] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@maddylewis] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: Not really a bug, this is a feature request.
- The offending PR has been commented on: n/a
- A discussion in #expensify-bugs has been started: I don't think this applies here
- Determine if we should create a regression test for this bug: Yes

___

**Regression Test Proposal**
1. Open App
2. Click on the Search icon
3. Type an invalid phone number e.g. `+444`
4. Verify the `Please enter a valid phone number...` message is shown below the input
5. Verify no results are returned
6. Type a valid phone number e.g. `+1 (500) 555-0007`
7. Verify no message is shown below the input
8. Verify the phone number / user is found

**Comment by MelvinBot** [comment]
⚠️ Looks like this issue was linked to a **Deploy Blocker** [here](https://github.com/Expensify/App/issues/18355)

If you are the assigned CME please investigate whether the linked PR caused a regression and leave a comment with the results.

If a regression has occurred and you are the assigned CM follow the instructions [here](https://stackoverflow.com/c/expensify/questions/13647).

If this regression could have been avoided please consider also proposing a recommendation to the PR checklist so that we can avoid it in the future.

**Comment by s77rt** [comment]
Checklist provided already https://github.com/Expensify/App/issues/17461#issuecomment-1532197564

**Comment by maddylewis** [comment]
what is going on with this issue!? let me hide duplicate comments and ill get the BZ checklist completed today.

**Comment by s77rt** [comment]
@maddylewis We had two PRs. So (I can't believe I'm saying this) but this is actually not Melvin's fault.

**Comment by maddylewis** [comment]
@Prince-Mendiratta - this is amazing!

**Comment by maddylewis** [comment]
### Payouts
- Reporting bonus: N/A (internal)
- Contributor: $1000 for 1st PR + $1000 for 2nd PR = $2000 total
- C+: $1000 for 1st PR + $1000 for 2nd PR = $2000 total

**Comment by maddylewis** [comment]
Offers sent!

**Comment by s77rt** [comment]
@maddylewis Payment is only due for one PR (both are basically the same)

**Comment by maddylewis** [comment]
paid!
